caches/infinispan/src/test/java/org/commonjava/maven/galley/cache/infinispan/AbstractFastLocalCacheBMUnitTest.java

```diff
@@ -0,0 +1,320 @@
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
+import org.junit.rules.TestName;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+public class AbstractFastLocalCacheBMUnitTest
+        extends CacheProviderTCK
+{
+    protected static EmbeddedCacheManager CACHE_MANAGER;
+
+    protected final PathGenerator pathgen = new HashedLocationPathGenerator();
+
+    protected final FileEventManager events = new TestFileEventManager();
+
+    protected final TransferDecorator decorator = new TestTransferDecorator();
+
+    protected final Executor executor = Executors.newFixedThreadPool( 5 );
+
+    @Rule
+    public TemporaryFolder temp = new TemporaryFolder();
+
+    @Rule
+    public TestName name = new TestName();
+
+    protected FastLocalCacheProvider provider;
+
+    protected Cache<String, String> cache = CACHE_MANAGER.getCache( NFSOwnerCacheProducer.CACHE_NAME );
+
+    protected String result;
+
+    @BeforeClass
+    public static void setupClass()
+    {
+        CACHE_MANAGER = new NFSOwnerCacheProducer().getCacheMgr();
+    }
+
+    @Before
+    public void setup()
+            throws Exception
+    {
+        final String nfsBasePath = createNFSBaseDir( temp.newFolder().getCanonicalPath() );
+        provider =
+                new FastLocalCacheProvider( new PartyLineCacheProvider( temp.newFolder(), pathgen, events, decorator ),
+                                            cache, pathgen, events, decorator, executor, nfsBasePath );
+        provider.init();
+    }
+
+    @After
+    public void tearDown()
+    {
+        provider.destroy();
+    }
+
+    @Override
+    protected FastLocalCacheProvider getCacheProvider()
+            throws Exception
+    {
+        return provider;
+    }
+
+    protected void latchWait( CountDownLatch latch )
+    {
+        try
+        {
+            latch.await();
+        }
+        catch ( InterruptedException e )
+        {
+            System.out.println( "Threads await Exception." );
+        }
+    }
+
+    private String createNFSBaseDir( String tempBaseDir )
+            throws IOException
+    {
+        File file = new File( tempBaseDir + "/mnt/nfs" );
+        file.delete();
+        file.mkdirs();
+        return file.getCanonicalPath();
+    }
+
+    protected class WriteLockThread
+            implements Runnable
+    {
+        private ConcreteResource res;
+
+        private CountDownLatch latch;
+
+        public WriteLockThread( ConcreteResource res, CountDownLatch latch )
+        {
+            this.res = res;
+            this.latch = latch;
+        }
+
+        @Override
+        public void run()
+        {
+            provider.lockWrite( res );
+            provider.waitForWriteUnlock( res );
```

Comment:
shouldn't we be trying to lockWrite() then unlockWrite()? Waiting for write unlock seems like something you do if it's already locked (or in case it's already locked). This part of the API is a little broken, admittedly...